libfwupdplugin/fu-device-retry.c

```diff
@@ -0,0 +1,182 @@
+
+	cond = g_new (FuDeviceRetryCondition, 1);
+	cond->domain = domain;
+	cond->code = code;
+	cond->reset_func = reset_func;
+	g_ptr_array_add (self->conditions, cond);
+}
+
+/**
+ * fu_device_retry_exec:
+ * @self: A #FuDeviceRetry
+ * @exec_func: (scope async): A function to execute
+ * @user_data: (nullable): a helper to pass to @user_data
+ * @error: A #GError
+ *
+ * Calls a specific function a number of times, optionally handling the error
+ * with a reset action.
+ *
+ * If fu_device_retry_add_condition() has not been used then all errors are
+ * considered non-fatal until the last try.
+ *
+ * If @reset_func returns %FALSE, then the function returns straight away
+ * without any pending retries.
+ *
+ * Since: 1.4.0
+ **/
+gboolean
+fu_device_retry_exec (FuDeviceRetry *self,
+		      FuDeviceRetryFunc exec_func,
+		      gpointer user_data,
+		      GError **error)
+{
+	g_return_val_if_fail (FU_IS_DEVICE_RETRY (self), FALSE);
+	g_return_val_if_fail (exec_func != NULL, FALSE);
+	g_return_val_if_fail (error != NULL, FALSE);
+
+	for (guint i = 0; ; i++) {
+		g_autoptr(GError) error_local =	NULL;
+
+		/* run function, if success return success */
+		if (exec_func (self->device, user_data, &error_local))
+			break;
+
+		/* sanity check */
+		if (error_local == NULL) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "exec failed but no error set!");
+			return FALSE;
+		}
+
+		/* too many retries */
+		if (i >= self->count - 1) {
+			g_propagate_prefixed_error (error,
+						    g_steal_pointer (&error_local),
+						    "failed after %u retries: ",
+						    self->count);
+			return FALSE;
+		}
+
+		/* show recoverable error on the console */
+		if (self->conditions->len == 0) {
+			g_debug ("failed on try %u of %u: %s",
+				 i + 1, self->count,
+				 error_local->message);
+			continue;
+		}
+
+		/* find the condition that matches */
+		for (guint j = 0; j < self->conditions->len; j++) {
+			FuDeviceRetryCondition *cond = g_ptr_array_index (self->conditions, j);
+			if (g_error_matches (error_local, cond->domain, cond->code)) {
+				if (!cond->reset_func (self->device, user_data, error))
+					return FALSE;
+			}
+		}
+	}
+
+	/* success */
+	return TRUE;
+
+}
+
+static void
+fu_device_retry_finalize (GObject *obj)
+{
+	FuDeviceRetry *self = FU_DEVICE_RETRY (obj);
+	g_ptr_array_unref (self->conditions);
+	G_OBJECT_CLASS (fu_device_retry_parent_class)->finalize (obj);
+}
+
+static void
+fu_device_retry_class_init (FuDeviceRetryClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	object_class->finalize = fu_device_retry_finalize;
+}
+
+static void
+fu_device_retry_init (FuDeviceRetry *self)
+{
+	self->conditions = g_ptr_array_new_with_free_func (g_free);
+}
+
+/**
+ * fu_device_retry_new:
+ * @device: A #FuDevice
+ * @count: The number of retries to attempt
+ *
+ * Creates a new retry object.
+ *
+ * No reference is taken on @device, and so it must remain valid for the
```

Comment:
Why not take the reference?  It seems like it would help avoid forgetting that this acts funny.

Reply:
I originally had this, but in the common case the subclassed FuDevice (e.g. FuCcgxHpiDevice) would put the FuDeviceRetry in the private data. This means the device refcount is 2. If the device goes away and we unref it, we drop the refcount to 1, but that doesn't finalize the subclassed object as there's already a "self" reference to itself in the FuDeviceRetry. This means we effectively leak the object.

Of course, the other thing to do would be to jam all the FuDeviceRetry functionality into FuDevice itself... but I didn't know if we wanted to blow up an already large object. I guess it's only really a guint and gpointer. Should I try that instead? It certainly makes all the refcount stuff easier.

I think I've convinced myself merging is indeed the right thing to do...

Reply:
Look across the tree and see what stuff uses retries.  Is it only USB?
Maybe it makes sense to let it live directly in the `FuUSbDevice` if it's only used there right now.  And if it needs to grow across buses add into FuDevice later.

Reply:
I want to use it in FuSynapticsMstDevice too, so I think FuDevice is the right place. Patch coming up....

Reply:
Okay, new much simpler version below.